python/ray/train/tests/test_mosaic_trainer.py

```diff
@@ -0,0 +1,138 @@
+import pytest
+
+import torch
+import torch.utils.data
+
+import torchvision
+from torchvision import transforms, datasets
+
+from torchmetrics.classification.accuracy import Accuracy
+from composer.core.evaluator import Evaluator
+from composer.models.tasks import ComposerClassifier
+import composer.optim
+from composer.loggers import InMemoryLogger
+from composer.algorithms import LabelSmoothing
+
+import ray
+from ray.air.config import ScalingConfig
+import ray.train as train
+from ray.air import session
+from ray.train.mosaic import MosaicTrainer
+
+
+mean = (0.507, 0.487, 0.441)
+std = (0.267, 0.256, 0.276)
+cifar10_transforms = transforms.Compose(
+    [transforms.ToTensor(), transforms.Normalize(mean, std)]
+)
+
+data_directory = "s3://air-example-data/cifar-10-python.tar.gz"
+train_dataset = torch.utils.data.Subset(
+    datasets.CIFAR10(
+        data_directory, train=True, download=True, transform=cifar10_transforms
+    ),
+    list(range(64)),
+)
+test_dataset = torch.utils.data.Subset(
+    datasets.CIFAR10(
+        data_directory, train=False, download=True, transform=cifar10_transforms
+    ),
+    list(range(64)),
+)
+
+scaling_config = ScalingConfig(num_workers=2, use_gpu=False)
+
+
+def trainer_init_per_worker(**config):
+    BATCH_SIZE = 32
+    # prepare the model for distributed training and wrap with ComposerClassifier for
+    # Composer Trainer compatibility
+    model = config.pop("model", torchvision.models.resnet18(num_classes=10))
+    model = ComposerClassifier(ray.train.torch.prepare_model(model))
+
+    # prepare train/test dataset
+    train_dataset = config.pop("train_dataset")
+    test_dataset = config.pop("test_dataset")
+
+    batch_size_per_worker = BATCH_SIZE // session.get_world_size()
+    train_dataloader = torch.utils.data.DataLoader(
+        train_dataset, batch_size=batch_size_per_worker, shuffle=True
+    )
+    test_dataloader = torch.utils.data.DataLoader(
+        test_dataset, batch_size=batch_size_per_worker, shuffle=True
+    )
+
+    train_dataloader = train.torch.prepare_data_loader(train_dataloader)
+    test_dataloader = train.torch.prepare_data_loader(test_dataloader)
+
+    evaluator = Evaluator(
+        dataloader=test_dataloader, label="my_evaluator", metrics=Accuracy()
+    )
+
+    # prepare optimizer
+    optimizer = composer.optim.DecoupledSGDW(
+        model.parameters(),
+        lr=0.05,
+        momentum=0.9,
+        weight_decay=2.0e-3,
+    )
+
+    if config.pop("eval", False):
+        config["eval_dataloader"] = evaluator
+
+    return composer.trainer.Trainer(
+        model=model, train_dataloader=train_dataloader, optimizers=optimizer, **config
+    )
+
```

Comment:
This example is very clean! Can we actually move this into a separate `mosaic_cifar_example.py` file in the ray/train/examples directory? Then we can import that example here to run as a test. This is similar to what we do for some of our existing examples: https://sourcegraph.com/github.com/ray-project/ray/-/blob/python/ray/train/tests/test_examples.py?L79:5

This way users have a standalone example they can look at and use, but we can also make sure it is tested in the CI.

Reply:
Could we also check that loss is decreasing like we do in [this test](https://sourcegraph.com/github.com/ray-project/ray/-/blob/python/ray/train/tests/test_torch_trainer.py?L48)?

Reply:
checking the decreasing loss would be checked only after the reporting is integrated, so will defer to future PR

Reply:
Sounds good!